job-server/src/main/scala/spark/jobserver/JobManager.scala

```diff
@@ -15,51 +15,39 @@ import scala.collection.JavaConverters._
- *
- * Each forked process has a working directory with log files for that context only, plus
- * a file "context.conf" which contains context-specific settings.
+ * SparkContext.  It is passed $deployMode $clusterAddr $actorName $systemConfigFile
  */
 object JobManager {
   val logger = LoggerFactory.getLogger(getClass)
 
   // Allow custom function to create ActorSystem.  An example of why this is useful:
   // we can have something that stores the ActorSystem so it could be shut down easily later.
-  // Args: workDir contextContent clusterAddress
-  def start(args: Array[String], makeSystem: Config => ActorSystem) {
-    val clusterAddress = AddressFromURIString.parse(args(2))
-    val workDir = Paths.get(args(0))
-    try {
-      if (!Files.exists(workDir)) {
-        System.err.println(s"WorkDir $workDir does not exist, creating.")
-        Files.createDirectories(workDir)
-      }
+  // Args: deployMode workDir clusterAddress systemConfig
+  // Allow custom function to wait for termination. Useful in tests.
+  def start(args: Array[String], makeSystem: Config => ActorSystem,
+            waitForTermination: (ActorSystem, String) => Unit) {
 
-      //Create context.conf in the work dir with the config string
-      Files.write(workDir.resolve("context.conf"),
-        Seq(args(1)).asJava,
-        Charset.forName("UTF-8"))
-      // Set system property LOG_DIR
-      System.setProperty("LOG_DIR", args(0))
-    } catch {
-      case e: IOException =>
-        System.err.println(s"Write context config into temp work directory $workDir, error: ${e.getMessage}")
-        sys.exit(1)
-    }
+    val deployMode = args(0)
+    val clusterAddress = AddressFromURIString.parse(args(1))
+    val managerName = args(2)
+    val systemConfigFile = new File(args(3))
 
-    val contextConfig = ConfigFactory.parseString(args(1))
-    val managerName = contextConfig.getString("context.actorname")
+    if (!systemConfigFile.exists()) {
+      System.err.println(s"Could not find system configuration file $systemConfigFile")
+      sys.exit(1)
+    }
 
     val defaultConfig = ConfigFactory.load()
-    val config = if (args.length > 3) {
-      val configFile = new File(args(3))
-      if (!configFile.exists()) {
-        System.err.println("Could not find configuration file " + configFile)
-        sys.exit(1)
-      }
-      ConfigFactory.parseFile(configFile).withFallback(defaultConfig)
+    val systemConfig = ConfigFactory.parseFile(systemConfigFile).withFallback(defaultConfig)
+    val config = if (deployMode == "cluster") {
+      logger.info("Cluster mode: Removing akka.remote.netty.tcp.hostname from config!")
+      logger.info("Cluster mode: Replacing spark.jobserver.sqldao.rootdir with container tmp dir.")
+      val sqlDaoDir = Files.createTempDirectory("sqldao")
+      val sqlDaoDirConfig = ConfigValueFactory.fromAnyRef(sqlDaoDir.toAbsolutePath.toString)
+      systemConfig.withoutPath("akka.remote.netty.tcp.hostname")
```

Comment:
Even though context creatio and job starting works but shouldn't we replace the hostname with the IP of current VM? In the logs of FrontEnd VM, we see `akka.tcp://JobServer@127.0.0.1:45012` which is misleading because instead of localhost it should be the IP of one of the slaves. Currently, we store the reference of slave in contextInitInfos hashmap in AkkaClusterSupervisorActor and using this reference we can communicate with the JobManagerActor but this hashmap is kindof temporary and will probably go away as we starting moving towards HA. @velvia @derSascha

Reply:
@bsikander You have to set `akka.remote.netty.tcp.hostname` in your config to the cluster side IP of your host running the web frontend. The code, mention above, runs inside the job manager (application master / context java vm). This process runs on another host than the web frontend in cluster setup and that's why hostname settings gets removed here.  Did you set `akka.remote.netty.tcp.hostname` in your config?

Reply:
Yes, I did set `akka.remote.netty.tcp.hostname` and I also understand that on the JobMangerActor side we remove this setting.

Here is the flow:
1- User sends POST /context to create a context
2- manager_start.sh is used to spawn a JVM which runs JobManager.
3- In cluster mode, JobManager loads the config file already placed in the Slave VM.
4- This code removes the `akka.remote.netty.hcp.hostname`.
5- Now, when JobManagerActor communicates back with web frontend the Akka address which web front receives is of the format `akka:tcp://JobServer@127.0.0.1:<port>`. This part is a little odd. Since this is a remote actor the address should be something like `akka:tcp://JobServer@10.250.x.x:<port>`. That is why I was asking that is this intentional and is the behavior correct?

Also, is is possible that instead of removing the hostname we replace the hostname with the current IP?

Reply:
@derSascha any opinion?

Reply:
Hmm.  My opinion is that 127.0.0.1 would not work in a real cluster, but what is everyone's testing results?  OTOH if you leave out the hostname Akka tries to get the hostname, so the question is what would u put there instead of what Akka guesses?

Reply:
@bsikander i can't reproduce this. Are you sure that you are not looking in some older logs? All my logs have something like `akka.tcp://JobServer@10.250.x.x`.

The basic idea is: Remove the hostname, set the cluster to join to `10.250.x.x` and now let akka choose the right network interface to connect to `10.250.x.x`.

Reply:
Yea, I see 127.0.0.1. Have you modified your /etc/hosts by any chance ?

I will try your new change today and will let you know.